tests/codegen.rs

```diff
@@ -1,118 +1,97 @@
-            .unwrap();
-    }
-
-    let url = format!("https://mkcert.org/generate/all/except/{except}");
-    eprintln!("fetching {url}...");
-    let body = reqwest::get(&url).await.unwrap().text().await.unwrap();
-    fs::write("fetched.pem", &body).unwrap();
-
-    // Split file contents into a Vec<Vec<&str>> where each inner Vec represents one certificate
-
-    let mut certs = Vec::with_capacity(64);
-    let mut current = Vec::with_capacity(16);
-    for ln in body.lines() {
-        if !ln.is_empty() {
-            current.push(ln);
-            continue;
-        } else if current.is_empty() {
-            continue;
+async fn new_generated_code_is_fresh() {
+    // Configure a Reqwest client that only trusts the CA certificate expected to be the
+    // root of trust for the CCADB server.
+    //
+    // If we see Unknown CA TLS validation failures from the Reqwest client in the future it
+    // likely indicates that the upstream service has changed certificate authorities. In this
+    // case the vendored root CA will need to be updated. You can find the current root in use with
+    // Chrome by:
+    //  1. Navigating to `https://ccadb-public.secure.force.com/mozilla/`
+    //  2. Clicking the lock icon.
+    //  3. Clicking "Connection is secure"
+    //  4. Clicking "Certificate is valid"
+    //  5. Clicking the "Details" tab.
+    //  6. Selecting the topmost "System Trust" entry.
+    //  7. Clicking "Export..."
+    //  8. Converting the downloaded PEM root to DER with:
+    //     `openssl x509 -inform pem -in <path> -outform der -out webpki-roots/tests/data/<root>.der`
+    //  9. Committing the updated .der root CA.
+    let root = include_bytes!("data/DigiCertGlobalRootCA.der");
+    let root = reqwest::Certificate::from_der(root).unwrap();
+    let client = reqwest::Client::builder()
+        .user_agent(format!("webpki-roots/v{}", env!("CARGO_PKG_VERSION")))
+        .add_root_certificate(root)
+        .build()
+        .unwrap();
+
+    let ccadb_url =
+        "https://ccadb-public.secure.force.com/mozilla/IncludedCACertificateReportPEMCSV";
+    eprintln!("fetching {ccadb_url}...");
+
+    let req = client.get(ccadb_url).build().unwrap();
+    let csv_data = client
+        .execute(req)
+        .await
+        .expect("failed to fetch CSV")
+        .text()
+        .await
+        .unwrap();
+
+    // Parse the CSV metadata.
+    let metadata = csv::ReaderBuilder::new()
+        .has_headers(true)
+        .from_reader(csv_data.as_bytes())
+        .into_deserialize::<CertificateMetadata>()
+        .collect::<Result<Vec<_>, _>>()
+        .unwrap();
+
+    // Filter for just roots with the TLS trust bit that are not distrusted as of today's date.
+    let trusted_tls_roots = metadata
+        .into_iter()
+        .filter(|root| root.trusted_for_tls(&Utc::now().naive_utc().date()))
+        .collect::<Vec<CertificateMetadata>>();
+
+    // Create an ordered BTreeMap of the roots, panicing for any duplicates.
```

Comment:
I think "panicing" is usually spelled "panicking"?

Reply:
:blush: good point. Fixed.